src/db.c

```diff
@@ -1152,7 +1164,7 @@ void moveCommand(client *c) {
         addReply(c,shared.czero);
         return;
     }
-    expire = getExpire(c->db,c->argv[1]);
+    expire = o->hasexpire ? getExpire(c->db,c->argv[1]) : -1;
```

Comment:
You're right, it's not needed, although it is very fast because it is already in the CPU cache here.

I suggest the getExpire can receive the value also so it can check the hasexpire flag. This makes it is hard call getExpire by mistake without first looking up the value in the db.

```C
long long getExpire(redisDb *db, robj *key, robj *val) {
    if (!val->hasexpire) return -1;
    ...
```

Reply:
Actually, this function is used in many places and it would be a too big change for this PR to add the `robj *val` parameter. (If I add it here, it should also be added in setExpire and removeExpire for symmetry, but setExpire looks up the main dict to reuse the key object, etc.) Instead of deleting the assert, I'm going to check the 'hasexpire' flag here.

Reply:
maybe we wanna re-consider.
when i reviewed removeExpire, i saw that you took the main dict's robj to turn off the expire bit, and was relieved to see that you didn't add a lookup, since it already had one, but that means we lose the opportunity to remove that excessive lookup.

there's also a lookup in setExpire that has always been there, and is needed in order to get the key name, but i suppose in many of the calls to these we already have the main dict's robj at hand.

maybe we don't want to do that in this PR, but rather in a followup one, but i think we may want to add the argument to these 3 functions.
or maybe run a benchmark to see if it's helpful, and drop that change if it doesn't.

p.s. this brings me to another thought which is the change in this PR getting mixed with the one about writable replicas (just because you moved the expiration check into the raw lookupKey.
on one hand, since we already know where we wanna get with the end result, it might be easier to mix it all in one PR so we can judge the final outcome. but on the other hand maybe we want separate PRs for separate changes (easier for release notes and maybe review), but then we need to decide on the order, and which portion of the change falls in each PR cleanly.

Reply:
If we implicitly require the key passed to setExpire to be the key in the main dict, then it is easy to make a mistake. Maybe we can pass the dictEntry of the main db dict to {set,get,remove}Expire instead? But the db API so far doesn't expose dict entries...

The only place where getExpire is called where the value hasn't been looked up before is in migrateCommand (cluster.c). In all other cases, i'm pretty sure the impact is not noticable, since the db dict allocations are in the CPU cache already = very fast.

Regarding the writable replicas, I haven't come to that yet, but if you decide exactly what you want, I can do it. :-)